### PR TITLE
Make Yarn container and helix instance allocation group by tag

### PR DESCRIPTION
**Pull request by hanghangliu**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-1661] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1661

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):
    This change enable Helix instances use the specific tag defined(if exists) in their Job/Workflow. This makes the Helix cluster management easier and more strait forward, as each job's instances can carry unique Helix tag. 

In this case, for each Gobblin job run, helix tasks within this job will
be assigned to containers with desired helix tag and resource. This can make within one Gobblin job run, multiple resource config can be set for different topics, instead of the need of creating new Helix cluster for different Gobblin job run setting.

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Passed all updated unit tests, tested in the cluster.

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-yarn/src/main/java/org/apache/gobblin/yarn/YarnService.java [hunk 1]
```diff
@@ -431,27 +427,35 @@ private EventSubmitter buildEventSubmitter() {
...
+      }
+      requestedContainerCountMap.put(currentHelixTag, desiredContainerCount);
+    }
+
+    // If a requested tag is not presented in the new request, update the requested count to 0 as we should release them
+    for(String requestedHelixTag : requestedContainerCountMap.keySet()) {
+      if(!yarnContainerRequestBundle.getHelixTagContainerCountMap().containsKey(requestedHelixTag)) {
+        requestedContainerCountMap.put(requestedHelixTag, 0);
+      }
     }
```
Comment (hanghangliu) [inline gobblin-yarn/src/main/java/org/apache/gobblin/yarn/YarnService.java:459]: This is the newly added fix for after PR 3487 and 3504

**Comment by hanghangliu** [comment]
This PR combines PR 3487 and 3504(which have been reverted) and new logic for the fix when a requested tag is not presented in the new request, requestedContainerCountMap is not update correctly.

File: gobblin-yarn/src/main/java/org/apache/gobblin/yarn/YarnService.java [hunk 2]
```diff
@@ -431,27 +427,35 @@ private EventSubmitter buildEventSubmitter() {
...
+      String currentHelixTag = entry.getKey();
+      int desiredContainerCount = entry.getValue();
+      int requestedContainerCount = requestedContainerCountMap.getOrDefault(currentHelixTag, 0);
+      for(; requestedContainerCount < desiredContainerCount; requestedContainerCount++) {
+        requestContainer(Optional.absent(), yarnContainerRequestBundle.getHelixTagResourceMap().get(currentHelixTag));
+      }
+      requestedContainerCountMap.put(currentHelixTag, desiredContainerCount);
+    }
+
+    // If a requested tag is not presented in the new request, update the requested count to 0 as we should release them
```
Comment (sv2000) [inline gobblin-yarn/src/main/java/org/apache/gobblin/yarn/YarnService.java:454]: Don't understand the comment. Are you suggesting setting the requested count to 0 to release containers? If so, why is that needed?

Comment (sv2000) [reply gobblin-yarn/src/main/java/org/apache/gobblin/yarn/YarnService.java:454]: Understood. Thanks for the clarification.

**Review by sv2000 — APPROVED** [review]
+1.
